File: backend-rust/src/util/app_token.rs
use crate::data::RecordManager;

/// Parsed metadata from decrypted app_token: `app_id_____login_____role_____access_right`
#[derive(Debug, Clone, Default)]
pub struct AppTokenMeta {
    pub app_id: String,
    pub login_identifier: String,
    pub role: String,
    pub access_right: i32,
}

pub fn parse_app_token(record_manager: &RecordManager, token: &str) -> AppTokenMeta {
    let token = token.trim();
    if token.is_empty() {
        return AppTokenMeta::default();
    }
    let decrypted = record_manager
        .csm_decrypt(token)
        .unwrap_or_else(|_| token.to_string());
    parse_decrypted_token(&decrypted)
}

pub fn parse_decrypted_token(decrypted: &str) -> AppTokenMeta {
    let parts: Vec<&str> = decrypted.split("_____").collect();
    AppTokenMeta {
        app_id: parts.first().copied().unwrap_or("").trim().to_string(),
        login_identifier: parts.get(1).copied().unwrap_or("").trim().to_string(),
        role: parts.get(2).copied().unwrap_or("").trim().to_string(),
        access_right: parts
            .last()
            .and_then(|s| s.trim().parse::<i32>().ok())
            .unwrap_or(0),
    }
}

pub fn is_sub_user_role(role: &str) -> bool {
    role.eq_ignore_ascii_case("user")
}

pub fn is_dev_access_right(access_right: i32) -> bool {
    access_right > 0
}

File: backend-go/internal/util/app_token.go
package util

import (
	"strings"

	"csm_server/backend-go/internal/data"
)

type AppTokenMeta struct {
	AppID            string
	LoginIdentifier  string
	Role             string
	AccessRight      int
}

func ParseAppToken(rm *data.RecordManager, token string) AppTokenMeta {
	token = strings.TrimSpace(token)
	if token == "" {
		return AppTokenMeta{}
	}
	decrypted, err := rm.CsmDecrypt(token)
	if err != nil {
		decrypted = token
	}
	return ParseDecryptedToken(decrypted)
}

func ParseDecryptedToken(decrypted string) AppTokenMeta {
	parts := strings.Split(decrypted, "_____")
	meta := AppTokenMeta{}
	if len(parts) > 0 {
		meta.AppID = strings.TrimSpace(parts[0])
	}
	if len(parts) > 1 {
		meta.LoginIdentifier = strings.TrimSpace(parts[1])
	}
	if len(parts) > 2 {
		meta.Role = strings.TrimSpace(parts[2])
	}
	if len(parts) > 0 {
		if v, ok := parseInt(parts[len(parts)-1]); ok {
			meta.AccessRight = v
		}
	}
	return meta
}

func IsSubUserRole(role string) bool {
	return strings.EqualFold(strings.TrimSpace(role), "user")
}

func AppIDFromToken(rm *data.RecordManager, token string) string {
	meta := ParseAppToken(rm, token)
	return meta.AppID
}

func parseInt(s string) (int, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, false
	}
	var n int
	for _, c := range s {
		if c < '0' || c > '9' {
			return 0, false
		}
		n = n*10 + int(c-'0')
	}
	return n, true
}
